rust-runtime/smithy-eventstream/src/frame.rs

```diff
@@ -0,0 +1,590 @@
+    const TYPE_FALSE: u8 = 1;
+    const TYPE_BYTE: u8 = 2;
+    const TYPE_INT16: u8 = 3;
+    const TYPE_INT32: u8 = 4;
+    const TYPE_INT64: u8 = 5;
+    const TYPE_BYTE_ARRAY: u8 = 6;
+    const TYPE_STRING: u8 = 7;
+    const TYPE_TIMESTAMP: u8 = 8;
+    const TYPE_UUID: u8 = 9;
+
+    /// Event Stream frame header value.
+    #[non_exhaustive]
+    #[derive(Clone, Debug, PartialEq)]
+    pub enum HeaderValue {
+        Bool(bool),
+        Byte(u8),
+        Int16(i16),
+        Int32(i32),
+        Int64(i64),
+        ByteArray(Bytes),
+        String(StrBytes),
+        Timestamp(Instant),
+        Uuid(u128),
+    }
+
+    macro_rules! read_value {
+        ($buf:ident, $typ:ident, $size_typ:ident, $read_fn:ident) => {
+            if $buf.remaining() >= size_of::<$size_typ>() {
+                Ok(HeaderValue::$typ($buf.$read_fn()))
+            } else {
+                Err(Error::InvalidHeaderValue)
+            }
+        };
+    }
+
+    impl HeaderValue {
+        pub(super) fn read_from(buffer: &mut dyn Buf) -> Result<HeaderValue, Error> {
+            let value_type = buffer.get_u8();
+            match value_type {
+                TYPE_TRUE => Ok(HeaderValue::Bool(true)),
+                TYPE_FALSE => Ok(HeaderValue::Bool(false)),
+                TYPE_BYTE => read_value!(buffer, Byte, u8, get_u8),
+                TYPE_INT16 => read_value!(buffer, Int16, i16, get_i16),
+                TYPE_INT32 => read_value!(buffer, Int32, i32, get_i32),
+                TYPE_INT64 => read_value!(buffer, Int64, i64, get_i64),
+                TYPE_BYTE_ARRAY | TYPE_STRING => {
+                    if buffer.remaining() > size_of::<u16>() {
+                        let len = buffer.get_u16() as usize;
+                        if buffer.remaining() < len {
+                            return Err(Error::InvalidHeaderValue);
+                        }
+                        let bytes = buffer.copy_to_bytes(len);
+                        if value_type == TYPE_STRING {
+                            Ok(HeaderValue::String(bytes.into()))
+                        } else {
+                            Ok(HeaderValue::ByteArray(bytes))
+                        }
+                    } else {
+                        Err(Error::InvalidHeaderValue)
+                    }
+                }
+                TYPE_TIMESTAMP => {
+                    if buffer.remaining() >= size_of::<i64>() {
+                        Ok(HeaderValue::Timestamp(Instant::from_epoch_seconds(
+                            buffer.get_i64(),
+                        )))
+                    } else {
+                        Err(Error::InvalidHeaderValue)
+                    }
+                }
+                TYPE_UUID => read_value!(buffer, Uuid, u128, get_u128),
+                _ => Err(Error::InvalidHeaderValueType(value_type)),
+            }
+        }
+
+        pub(super) fn write_to(&self, buffer: &mut dyn BufMut) -> Result<(), Error> {
```

Comment:
I think I'd probably use `impl BufMut` here? although I'm generally pro dyn, these methods are going to be extremely hot so getting it to be fully optimized is maybe worth it?

Reply:
discussion of the tradeoffs here: https://users.rust-lang.org/t/should-i-prefer-impl-impl-mut-impl-or-box-dyn-dyn-mut-dyn-by-default/41778/9